File: app/src/main/java/com/cvv/fanstaticapps/randomticker/activities/KlaxonActivity.java
package com.cvv.fanstaticapps.randomticker.activities;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.content.Intent;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.CycleInterpolator;
import android.view.animation.RotateAnimation;

import com.cvv.fanstaticapps.randomticker.R;
import com.cvv.fanstaticapps.randomticker.helper.TimerHelper;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;
import io.github.kobakei.grenade.annotation.Extra;
import io.github.kobakei.grenade.annotation.Navigator;
import pl.bclogic.pulsator4droid.library.PulsatorLayout;

import static com.cvv.fanstaticapps.randomticker.helper.TimerHelper.ONE_SECOND_IN_MILLIS;

@Navigator
public class KlaxonActivity extends BaseActivity {

    private static final int ANIMATION_DURATION = 750;
    private static final String TAG = KlaxonActivity.class.getSimpleName();

    @Extra
    boolean timeElapsed;

    @BindView(R.id.waiting_icon)
    View waitingIcon;
    @BindView(R.id.timer_hand)
    View timerHand;
    @BindView(R.id.root)
    View root;
    @BindView(R.id.dismiss_button)
    View dismissButton;
    @BindView(R.id.pulsator)
    PulsatorLayout pulsatorLayout;

    @Inject
    TimerHelper timerHelper;

    private Ringtone playingAlarmSound;
    private Animation waitingIconAnimation;
    //timestamp when the timer should ring
    private long intervalFinished;
    private CountDownTimer countDownTimer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_klaxon);
        KlaxonActivityNavigator.inject(this, getIntent());

        intervalFinished = preferences.getIntervalFinished();
    }

    @Override
    protected void onPostCreate(@Nullable Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        if (timeElapsed) {
            timerFinished();
        } else {
            startCountDownTimer();
        }
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        KlaxonActivityNavigator.inject(this, intent);
        if (timeElapsed && countDownTimer != null) {
            countDownTimer.cancel();
            timerFinished();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        cancelEverything();
    }

    private void startCountDownTimer() {
        countDownTimer = new CountDownTimer(intervalFinished - System.currentTimeMillis(), ONE_SECOND_IN_MILLIS) {
            @Override
            public void onTick(long millisUntilFinished) {
                //nothing to do
            }

            @Override
            public void onFinish() {
                timerFinished();
                preferences.setCurrentlyTickerRunning(false);
                countDownTimer = null;
            }
        };
        countDownTimer.start();
        startBellAnimation();
    }

    private void timerFinished() {
        timerHelper.cancelNotification(this, preferences);
        playRingtone();
        if (!pulsatorLayout.isStarted()) {
            hideBellAndMoveCancelButton();
        }
    }

    private void hideBellAndMoveCancelButton() {
        waitingIconAnimation.cancel();
        waitingIcon.animate()
                .alpha(0)
                .scaleX(0)
                .scaleY(0)
                .setDuration(ANIMATION_DURATION)
                .setListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        super.onAnimationEnd(animation);
                        waitingIcon.setVisibility(View.GONE);
                    }
                }).start();

        pulsatorLayout.animate()
                .scaleX(1.5f)
                .setStartDelay(100)
                .scaleY(1.5f)
                .setListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        super.onAnimationEnd(animation);
                        pulsatorLayout.start();
                    }
                })
                .setDuration(ANIMATION_DURATION);
    }

    @OnClick(R.id.dismiss_button)
    void onDismissClicked() {
        cancelEverything();
        timerHelper.cancelNotificationAndGoBack(this, preferences);
    }

    private void cancelEverything() {
        if (playingAlarmSound != null) {
            playingAlarmSound.stop();
        }
        if (waitingIconAnimation != null) {
            waitingIconAnimation.cancel();
        }
        if (countDownTimer != null) {
            countDownTimer.cancel();
        }
    }

    private void startBellAnimation() {
        waitingIconAnimation = new RotateAnimation(0, 360, Animation.RELATIVE_TO_SELF,
                0.5f, Animation.RELATIVE_TO_SELF, 0.55f);
        waitingIconAnimation.setRepeatCount(Animation.INFINITE);
        waitingIconAnimation.setInterpolator(new CycleInterpolator(1));
        waitingIconAnimation.setDuration(4000);
        timerHand.startAnimation(waitingIconAnimation);
    }


    private void playRingtone() {
        if (playingAlarmSound == null) {
            try {
                String preferenceRingtone = PreferenceManager
                        .getDefaultSharedPreferences(this)
                        .getString(getString(R.string.pref_ringtone),
                                RingtoneManager.getDefaultUri(RingtoneManager.TYPE_ALARM).toString());
                Uri uri = Uri.parse(preferenceRingtone);
                playingAlarmSound = RingtoneManager.getRingtone(getApplicationContext(), uri);
                playingAlarmSound.play();
            } catch (Exception e) {
                Log.e(TAG, "Error while trying to play alarm sound", e);
            }
        } else {
            toast(R.string.bell_ringing);
        }
    }

}

File: app/src/main/java/com/cvv/fanstaticapps/randomticker/activities/KlaxonActivity.kt
package com.cvv.fanstaticapps.randomticker.activities

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import android.content.Context
import android.content.Intent
import android.media.Ringtone
import android.media.RingtoneManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import android.view.View
import android.view.WindowManager
import android.view.animation.Animation
import android.view.animation.CycleInterpolator
import android.view.animation.RotateAnimation
import com.cvv.fanstaticapps.randomticker.PREFS
import com.cvv.fanstaticapps.randomticker.R
import com.cvv.fanstaticapps.randomticker.helper.TimerHelper
import com.cvv.fanstaticapps.randomticker.helper.WakeLocker
import com.cvv.fanstaticapps.randomticker.mvp.KlaxonPresenter
import com.cvv.fanstaticapps.randomticker.mvp.KlaxonPresenter.ViewState
import com.cvv.fanstaticapps.randomticker.mvp.KlaxonView
import io.github.kobakei.grenade.annotation.Extra
import io.github.kobakei.grenade.annotation.Navigator
import kotlinx.android.synthetic.main.activity_klaxon.*
import timber.log.Timber
import javax.inject.Inject


@Navigator
class KlaxonActivity : BaseActivity(), KlaxonView {

    private val animationDuration = 750

    @JvmField
    @Extra
    var timeElapsed: Boolean = false

    @Inject
    lateinit var timerHelper: TimerHelper

    private var playingAlarmSound: Ringtone? = null
    private var waitingIconAnimation: Animation? = null
    private var vibrator: Vibrator? = null

    private lateinit var presenter: KlaxonPresenter;
    private var elapsedTimeNeedsAnimation: Boolean = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WakeLocker.acquireLock(this)

        setContentView(R.layout.activity_klaxon)

        enableShowWhenLocked()

        presenter = KlaxonPresenter(this, PREFS.intervalFinished, timeElapsed)

        KlaxonActivityNavigator.inject(this, intent)
    }

    @Suppress("DEPRECATION")
    private fun enableShowWhenLocked() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            setShowWhenLocked(true)
            setTurnScreenOn(true)
        } else {
            window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED
                    or WindowManager.LayoutParams.FLAG_DISMISS_KEYGUARD
                    or WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
                    or WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON
                    or WindowManager.LayoutParams.FLAG_ALLOW_LOCK_WHILE_SCREEN_ON)
        }
    }

    override fun onPause() {
        super.onPause()
        WakeLocker.release()
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        dismissButton.setOnClickListener {
            presenter.cancelTimer()
        }

        presenter.init()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        KlaxonActivityNavigator.inject(this, intent)
        presenter.update(timeElapsed)
    }

    override fun onStop() {
        super.onStop()
        presenter.onStop()
    }

    override fun render(viewState: ViewState) {
        when (viewState) {
            is ViewState.TimerStarted -> {
                startBellAnimation()
                elapsedTime.setOnClickListener { presenter.showElapsedTime = true }

            }
            is ViewState.ElapseTimeUpdate -> {
                elapsedTime.text = viewState.elapsedTime
            }
            is ViewState.TimerFinished -> {
                timerHelper.cancelNotification(this, PREFS)
                showElapsedTime(viewState.elapsedTime)
                playRingtone()
                vibrate()
                if (!pulsator.isStarted) {
                    hideWaitingIcon()
                    startPulsatorAnimation()
                }
            }
            is ViewState.TimerCanceled -> {
                cancelEverything()
                timerHelper.cancelNotificationAndGoBack(this, PREFS)
            }
            is ViewState.TimerStopped -> {
                cancelEverything()
            }
        }
    }

    private fun cancelEverything() {
        playingAlarmSound?.stop()
        waitingIconAnimation?.cancel()
        vibrator?.cancel()
    }

    private fun showElapsedTime(elapsedTimeInMillis: String) {
        elapsedTime.text = elapsedTimeInMillis
        if (elapsedTimeNeedsAnimation) {
            elapsedTimeNeedsAnimation = false
            val startSize = resources.getDimensionPixelSize(R.dimen.elepsedTimeTextSize).toFloat()
            val endSize = resources.getDimensionPixelSize(R.dimen.elepsedTimeTextSizeZoomed).toFloat()

            val animator = ValueAnimator.ofFloat(startSize, endSize)
            animator.duration = 1200


            animator.addUpdateListener { valueAnimator ->
                val animatedValue = valueAnimator.animatedValue as Float
                elapsedTime.setTextSize(animatedValue)
            }

            animator.start()
        }
    }

    private fun hideWaitingIcon() {
        waitingIconAnimation?.cancel()
        waitingIcon.animate()
                .alpha(0f)
                .scaleX(0f)
                .scaleY(0f)
                .setDuration(animationDuration.toLong())
                .setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator) {
                        super.onAnimationEnd(animation)
                        waitingIcon!!.visibility = View.GONE
                    }
                }).start()
    }

    private fun startPulsatorAnimation() {
        pulsator.animate()
                .scaleX(1.5f)
                .setStartDelay(100)
                .scaleY(1.5f)
                .setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationEnd(animation: Animator) {
                        super.onAnimationEnd(animation)
                        pulsator.start()
                    }
                }).duration = animationDuration.toLong()
    }

    private fun startBellAnimation() {
        waitingIconAnimation = RotateAnimation(0f, 360f, Animation.RELATIVE_TO_SELF,
                0.5f, Animation.RELATIVE_TO_SELF, 0.55f)
        waitingIconAnimation!!.repeatCount = Animation.INFINITE
        waitingIconAnimation!!.interpolator = CycleInterpolator(1f)
        waitingIconAnimation!!.duration = 4000
        timerHand.startAnimation(waitingIconAnimation)
    }

    private fun playRingtone() {
        if (playingAlarmSound == null) {
            try {
                val uri = Uri.parse(PREFS.alarmRingtone)
                playingAlarmSound = RingtoneManager.getRingtone(applicationContext, uri)
                playingAlarmSound!!.play()
            } catch (e: Exception) {
                Timber.e(e, "Error while trying to play alarm sound")
            }
        } else {
            toast(R.string.bell_ringing)
        }
    }

    @Suppress("DEPRECATION")
    private fun vibrate() {
        if (vibrator == null && PREFS.vibrator) {
            vibrator = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
            if (Build.VERSION.SDK_INT >= 26) {
                vibrator?.vibrate(VibrationEffect.createOneShot(230, VibrationEffect.DEFAULT_AMPLITUDE))
            } else {
                vibrator?.vibrate(230);
            }
        }
    }

}